editor/components/post-permalink/index.js

```diff
@@ -14,55 +14,151 @@ import { Dashicon, ClipboardButton, Button } from '@wordpress/components';
  * Internal Dependencies
  */
 import './style.scss';
-import { isEditedPostNew, getEditedPostAttribute } from '../../store/selectors';
+import { isEditedPostNew, getEditedPostAttribute, getEditedPostPreviewLink } from '../../store/selectors';
+import { editPost } from '../../store/actions';
+import { getWPAdminURL } from '../../utils/url';
 
 class PostPermalink extends Component {
 	constructor() {
 		super( ...arguments );
+
 		this.state = {
-			showCopyConfirmation: false,
+			iconClass: '',
+			samplePermalink: '',
+			samplePermalinkPrefix: '',
+			samplePermalinkSuffix: '',
+			postName: '',
+			editedPostName: '',
+			editingPermalink: false,
 		};
-		this.onCopy = this.onCopy.bind( this );
-		this.onFinishCopy = this.onFinishCopy.bind( this );
+
+		this.onSavePermalink = this.onSavePermalink.bind( this );
 	}
 
-	componentWillUnmount() {
-		clearTimeout( this.dismissCopyConfirmation );
+	componentWillMount() {
+		if ( this.props.samplePermalinkData ) {
+			this.updateSamplePermalink( this.props.samplePermalinkData[ 0 ], this.props.samplePermalinkData[ 1 ] );
```

Comment:
This is a once-documented premature optimization anti-pattern, as it's duplicating the source of truth, not necessarily state of the component:

https://web.archive.org/web/20160531222015/http://facebook.github.io:80/react/tips/props-in-getInitialState-as-anti-pattern.html

It will save many a headache to simply calculate the value as used.

Reply:
Further, a better optimization we might consider is to create a separate component which manages the editing behaviors, and is only rendered when `this.state.editingPermalink` is true.

(i.e. why do we calculate this state when I'm merely editing the title's text, or viewing the permalink?)

Reply:
Good point, changed to just calculating when rendering the form.